src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -786,38 +836,42 @@ evalApp γ ctx e0 es et
   , Just eq <- Map.lookup f (knAms γ)
   , length (eqArgs eq) <= length es
   = do
-       env  <- gets (seSort . evEnv)
+       env <- gets (seSort . evEnv)
        okFuel <- checkFuel f
-       if okFuel && et /= FuncNormal
-         then do
-                let (es1,es2) = splitAt (length (eqArgs eq)) es
-                    newE = substEq env eq es1
-                (e', fe) <- evalIte γ ctx et newE -- TODO:FUEL this is where an "unfolding" happens, CHECK/BUMP counter
-                let e2' = stripPLEUnfold e'
-                    e3' = simplify γ ctx (eApps e2' es2) -- reduces a bit the equations
-                    undecidedGuards = case e' of
-                      EIte{} -> True
-                      _ -> False
-
-                if undecidedGuards
-                  then do
-                    modify $ \st ->
-                      st {
-                        evPendingUnfoldings = M.insert (eApps e0 es) e3' (evPendingUnfoldings st)
-                      }
-                    -- Don't unfold the expression if there is an if-then-else
-                    -- guarding it, just to preserve the size of further
-                    -- rewrites.
-                    return (Nothing, noExpand)
-                  else do
-                    useFuel f
-                    modify $ \st ->
-                      st
-                        { evNewEqualities = S.insert (eApps e0 es, e3') (evNewEqualities st)
-                        , evPendingUnfoldings = M.delete (eApps e0 es) (evPendingUnfoldings st)
-                        }
-                    return (Just $ eApps e2' es2, fe)
-         else return (Nothing, noExpand)
+       if okFuel && et /= FuncNormal then do
+         let (es1, es2) = splitAt (length (eqArgs eq)) es
+         -- We need the function with ECst annotations since they
+         -- are needed for doing eta expansion, doing elaboration
+         -- of the function body ahead of time would cause too much
+         -- monomorphization on the LH side, suppose we have a
+         -- polymorphic function of Haskell type `forall a. (a -> a)`,
+         -- we want the type `forall a. (ECst a a -> ECst a a)`
+         -- if we perform the elaboration ahead of time we would
+         -- get a term annoptated with type `fun(0, [a#foo, a#foo])`
+         -- instead of `fun(0, [@foo, @foo])` thus when substituting
+         -- the body in a term we could encounter type mismatches.
+         newE <- elaborateExpr "EvalApp unfold full: " $ substEq env eq es1
```

Comment:
If this had an effect on the performance of the tests that are slowing down, perhaps elaboration could be done only if `--etabeta` is enabled.